ebpf/python/pyperf.go

```diff
@@ -0,0 +1,264 @@
+	"github.com/grafana/pyroscope/ebpf/metrics"
+	"github.com/grafana/pyroscope/ebpf/symtab"
+	lru "github.com/hashicorp/golang-lru/v2"
+)
+
+type Perf struct {
+	rd             *perf.Reader
+	logger         log.Logger
+	pidDataHashMap *ebpf.Map
+	symbolsHashMp  *ebpf.Map
+	metrics        *metrics.PythonMetrics
+
+	events      []*PerfPyEvent
+	eventsLock  sync.Mutex
+	sc          *symtab.SymbolCache
+	pidCache    *lru.Cache[uint32, *PerfPyPidData]
+	prevSymbols map[uint32]*PerfPySymbol
+	wg          sync.WaitGroup
+}
+
+func NewPerf(logger log.Logger, metrics *metrics.PythonMetrics, perfEventMap *ebpf.Map, pidDataHasMap *ebpf.Map, symbolsHashMap *ebpf.Map) (*Perf, error) {
+	rd, err := perf.NewReader(perfEventMap, 4*os.Getpagesize())
+	if err != nil {
+		return nil, fmt.Errorf("perf new reader: %w", err)
+	}
+	pidCache, err := lru.NewWithEvict[uint32, *PerfPyPidData](512, func(key uint32, value *PerfPyPidData) {
+		_ = pidDataHasMap.Delete(key)
+	})
+	if err != nil {
+		return nil, fmt.Errorf("pyperf pid cache %w", err)
+	}
+	res := &Perf{
+		rd:             rd,
+		logger:         logger,
+		pidDataHashMap: pidDataHasMap,
+		symbolsHashMp:  symbolsHashMap,
+		pidCache:       pidCache,
+		metrics:        metrics,
+	}
+	res.wg.Add(1)
+	go func() {
+		defer res.wg.Done()
+		res.loop()
+	}()
+	return res, nil
+}
+
+// StartPythonProfiling return python init error, second return value is true if process is still alive,
+// third return value - python process data
+func (s *Perf) StartPythonProfiling(pid uint32, data *PerfPyPidData, serviceName string) error {
+	if s.pidCache.Contains(pid) {
+		return nil
+	}
+
+	err := s.pidDataHashMap.Update(pid, data, ebpf.UpdateAny)
+	if err != nil { // should never happen
+		return fmt.Errorf("updating pid data hash map: %w", err)
+	}
+	s.metrics.ProcessInitSuccess.WithLabelValues(serviceName).Inc()
+	s.pidCache.Add(pid, data)
+	return nil
+}
+
+func (s *Perf) loop() {
+	defer s.rd.Close()
+
+	for {
+		record, err := s.rd.Read()
+		if err != nil {
+			if errors.Is(err, perf.ErrClosed) {
+				return
+			}
+			_ = level.Error(s.logger).Log("msg", "[pyperf] reading from perf event reader", "err", err)
+			continue
+		}
+
+		if record.LostSamples != 0 {
+			s.metrics.LostSamples.Add(float64(record.LostSamples))
+			_ = level.Debug(s.logger).Log("msg", "[pyperf] perf event ring buffer full, dropped samples", "n", record.LostSamples)
+		}
+
+		if record.RawSample != nil {
+			event, err := ReadPyEvent(record.RawSample)
+			if err != nil {
+				_ = level.Error(s.logger).Log("msg", "[pyperf] parsing perf event record", "err", err)
+				continue
+			}
+			s.eventsLock.Lock()
+			s.events = append(s.events, event)
+			s.eventsLock.Unlock()
+		}
+	}
+
+}
+
+func (s *Perf) Close() {
+	_ = s.rd.Close()
+	s.wg.Wait()
+}
+
+func (s *Perf) ResetEvents() []*PerfPyEvent {
+	s.eventsLock.Lock()
+	defer s.eventsLock.Unlock()
+	if len(s.events) == 0 {
+		return nil
+	}
+	eventsCopy := make([]*PerfPyEvent, len(s.events))
```

Comment:
the reason I chose to use a pointer is because PerfPyEvent is quite big - 320 bytes
and there could be quite a lot of them 100 times per second * 15 seconds * 16 cores * 320 bytes =~7M
And i did not want to move this memory twice. It is likely a premature optimization already. At the same time I dont see this function hot, so I suggest we keep it as is and optimize if we have a problem or any evidence it can be optimized